### ISSUE TITLE
Feature Request: Fuzzy Matching and/or Partial Completion

### ISSUE THREAD

**Issue by lmburns** [Jarsop/rualdi#5]
I really like the idea how this works compared to `zoxide`. That instead of marking every directory you enter, you're able to specify exactly what directories you would like to bookmark (somewhat). 

I was wondering if a feature could be implemented to fuzzy complete the aliases when using them. This could be implemented in a couple of ways. 
* A function could be called that invokes `fzf` for the user to choose the alias and switch directories
* Dynamic `zsh` completions that read the file in which the aliases are stored, then the user could press tab to complete the alias (also, for a bonus the user could use `fzf-tab` to use `fzf` with the completions, though this has to be with `zsh`)

Another feature that I would like to see implemented that is also kind of implemented in `zoxide` is that if I were to have only one alias that starts with a `c`, I could type `rad c` and it would use that directory. If I had two aliases that both start with a 'c', one 'curr', another 'crow', then typing `cr`, `cro`, or `crow` would all complete the `crow` alias.

Lastly, unrelated to this feature request, what exactly are the environment variables supposed to be used for? When I define an environment variable and trying printing it out to the console, nothing is shown. Is it supposed to be a feature where you can `cd` using the environment variable as well as the alias?

Edit: Also maybe add a slight bit of color to output if possible?
I'm working on `zsh` completions right now.

Edit again:
Here are some extremely rudimentary `zsh` completions if they were to be of any help, I know they are for me. I did not add Windows support, and this program requires `dasel` to parse the `toml` file. There is another program out there called `tomlq`, but `dasel` is probably used much more. Parsing a `toml` file is possible with pure `zsh`, though it would require a little more work.

```zsh
#compdef __rualdi_rad __rualdi_radr __rualdi_radx __rualdi_radxr

local -a aliases envs
local conf state ret
[[ "$OSTYPE" = darwin* ]] && conf="$HOME/Library/Application Support/rualdi/rualdi.toml"
[[ "$OSTYPE" = *linux* ]] && conf="$XDG_CONFIG_HOME/rualdi/rualdi.toml"

aliases=( ${(f)"$(dasel select -f "$conf" -m '.aliases.-')"} )
envs=( ${(f)"$(dasel select -f "$conf" -m '.environment.-')"} )

case $state in
  *)
    _alternative \
      'aliases:aliases:compadd -a - aliases' \
      'envs:environments:compadd -a - envs' && \
          ret=0
  ;;
esac

return ret
```

**Comment by lmburns** [comment]
Also, here is a function that is modified from your `zsh` init script and `wfxr/ormarks`.

```zsh
# Desc: fzf rualdi (dir aliases)

setopt extendedglob noshortloops rcexpandparam
zmodload -Fa zsh/parameter p:commands

local conf
local -a aliases

# DARWIN_CONF_HOME is set to $HOME/Library/Application Support
conf="${${(b)DARWIN_CONF_HOME}:-$XDG_CONFIG_HOME}/rualdi/rualdi.toml"
aliases=( ${${(f)"$(dasel select -f "$conf" '.aliases')"}//\"} )

# Selects only aliases, no paths
# aliases=( ${(f)"$(dasel select -f "$conf" -m '.aliases.-')"} )

# Modified from rualdi
function __fk_cd() {
  local cpwd
  builtin cd "$@" || return "$?"
  cpwd=$(builtin pwd -L)
  # (D) will replace expanded $HOME with ~
  builtin print -Pr "%F{11}====%f %F{14}%B${(D)cpwd}%f%b %F{11}=====%f"
}

# Modified from wfxr/formarks
function __fk_colorize() {
  local field='\(\S\+\s*\)'
  local esc=$(printf '\033')
  local N="${esc}[0m"
  local R="${esc}[31m"
  local G="${esc}[32m"
  local Y="${esc}[33m"
  local B="${esc}[34m"
  local pattern="s#^${field}${field}${field}#$Y\1$R\2$N$B\3$N#"
  (( $+commands[gsed] )) && gsed "$pattern" || sed "$pattern"
}

function __fk_list() {
  # TODO: Add support for column -t file names with spaces
  aliases=( ${aliases// = / -> } )
  builtin print -rl -- "$aliases[@]" \
    | __fk_colorize \
    | column -t
}

function __fk_rad() {
  if [[ $# -eq 1 ]] && [[ "$1" = '-' ]]; then
    if [[ -n "$OLDPWD" ]]; then
      __fk_cd "$OLDPWD"
    else
      builtin print -Pr "%F{3}rualdi: \$OLDPWD is not set"
      return 1
    fi
  else
    local sel __fk_result

    sel=${${(@s:->:)$(__fk_list | __fk_colorize | fzf +m)}[1]}

    [[ -n "$sel" ]] &&
      __fk_cd "$(rualdi resolve -- "$sel")"
  fi
}

__fk_rad "$@"

# vim: ft=zsh:et:sw=0:ts=2:sts=2:fdm=marker:fmr={{{,}}}:
```

Adding a modification to printing the directory just because
```zsh
function __fk_cd() {
  local cpwd colp
  builtin cd "$@" || return "$?"
  cpwd=$(builtin pwd -L)
  # (D) will replace expanded $HOME with ~
  cpwd=${(D)cpwd}
  colp=${(l:(COLUMNS-$#cpwd-2)/2::=:):-}
  builtin print -Pr "%F{11}$colp%f %F{14}%B$cpwd%f%b %F{11}$colp%f"
}
```

**Comment by Jarsop** [comment]
@lmburns Thanks for your feedback and your proposal! It's really appreciated. I will work on maybe during this week, but I have other stuff to done first.
Don't hesitate to contribute via PR if you want 😁. You seem like a ZSH guru 😉. Have a good day and sorry for the delay

**Comment by Jarsop** [comment]
> Here are some extremely rudimentary zsh completions if they were to be of any help, I know they are for me. I did not add Windows support, and this program requires dasel to parse the toml file. There is another program out there called tomlq, but dasel is probably used much more. Parsing a toml file is possible with pure zsh, though it would require a little more work.

Maybe it's better to create a command in `rualdi` like `rualdi show [bash|zsh]` to format the TOML to the wanted output. It's avoid to create a second logic externally

**Comment by Jarsop** [comment]
I also saw your fork quickly, you have some good ideas with coloring output. Can you create a proposal to integrate it in the next version ?

**Comment by lmburns** [comment]
Yes, I can do that. I'm not sure exactly where you want to go with the project. I'd say I probably went overboard in the formatting when listing the aliases. I created a hash that will replace common directory prefixes with a percent sign and a name. You probably do not want that, and I'm not sure that other people would either. It just makes the paths a little shorter and I had done the exact same thing with a `zsh` program that I've wrote.

I think I'm going to try and implement something where the user could set a couple of colors within the configuration file to display the output. I'm very new to Rust and decided to fork this project for me to have something to work on to try and learn the language. I was also trying to make this similar to [`formarks`](https://github.com/wfxr/formarks) as mentioned above, but with a couple more features.

I added an `asciinema` video to my fork if you want to look at it to see what it is like. Let me know exactly what you'd like to be done and I can try and work on it. Also, are there any other features that are not implemented now that you'd like to implement? I added completions, to where (at least for `zsh`) you are able to hit tab and complete the directories when using the `rad` command.
